msprime/simulations.py

```diff
@@ -2355,6 +2357,263 @@ def population_size_trajectory(self, steps):
+        """
+        Returns an array such that P[j, a, b] is the probability that a lineage that
+        started in population a at time sample_time is in population b at time steps[j]
+        ago.
+
+        This function reports sampling probabilities _before_ mass migration events
+        at a step time, if a mass migration event occurs at one of those times.
+        Migrations will then effect then next time step
+
+        :param list steps: A list of times to compute probabilities of lineages.
+        :param sample_time: The time of sampling of the lineage. For any times in steps
+            that are more recent than sample_time, the probability of finding the
+            lineage in any population is zero.
+        :return: An array of dimension len(steps) by num pops by num_pops.
+        """
+        num_pops = self.num_populations
+        # P[i, j] will be the probability that a lineage that started in i is now in j
+        P = np.zeros([num_pops, num_pops])
+
+        for ii in range(num_pops):
+            P[ii, ii] = 1
+
+        # epochs are defined by mass migration events or changes to population sizes
+        # or migration rates, so we add the epoch interval times to the steps that we
+        # need to account for
+        epoch_breaks = [t for t in self.epoch_times if t not in steps]
+        all_steps = np.concatenate([steps, epoch_breaks])
+
+        # add sample time if not in epoch_breaks
+        sampling = []
+        if sample_time not in all_steps:
+            sampling.append(sample_time)
+        all_steps = np.concatenate((all_steps, sampling))
+
+        ix = np.argsort(all_steps)
+        all_steps = all_steps[ix]
+        # keep track of the steps to report in P_out
+        keep_steps = np.concatenate(
+            [
+                np.repeat(True, len(steps)),
+                np.repeat(False, len(epoch_breaks)),
+                np.repeat(False, len(sampling)),
+            ]
+        )[ix]
+
+        assert len(np.unique(all_steps)) == len(all_steps)
+        assert np.all(steps == all_steps[keep_steps])
+        P_out = np.zeros((len(all_steps), num_pops, num_pops))
+
+        first_step = 0
+        while all_steps[first_step] < sample_time:
+            first_step += 1
+
+        P_out[first_step] = P
+
+        # get ordered mass migration events
+        mass_migration_objects = []
+        mass_migration_times = []
+        for demo in self.demographic_events:
+            if demo.type == "mass_migration":
+                mass_migration_objects.append(demo)
+                mass_migration_times.append(demo.time)
+
+        for jj in range(first_step, len(all_steps) - 1):
+            t_j = all_steps[jj]
+
+            # apply any mass migration events to P
+            # so if we sample at this time, we do no account for the instantaneous
+            # mass migration events that occur at the same time. that will show up
+            # at the next step
+            if t_j > sample_time:
```

Comment:
Isn't this always true since we started at `first_step`?

Reply:
I want to exclude the migration events that occur at the same time as the sampling time.